### ISSUE TITLE
PDF desktop search crashes since version 3.3

### ISSUE THREAD

**Issue by sschuberth** [sumatrapdfreader/sumatrapdf#2012]
I've enabled the Windows Desktop Search option in the installer to let Windows Explorer (or in my case Directory Opus) search PDF file contents. This used to work fine in Sumatra PDF 3.2, but since the upgrade to version 3.3 the same search takes down Directory Opus and crashes.

I was looking at https://www.sumatrapdfreader.org/docs/Debugging-Sumatra to get debug logs, but those instructions do only seem to apply to the Sumatra main executable, not to the desktop search. Any hints where I can get logs for the desktop search?

**Comment by kjk** [comment]
I'm not sure it's possible, because it's a dll being executed inside windows process. I wouldn't know how to debug.

At the same time I don't see why it should crash Directory Opus. Is Directory Opus calling PdfFilter.dll directly?

I guess you could attach WinDBG to Directory Opus and if it crashes, you should at least be able to see where it crashed and if PdfFilter.dll is on the crash callstack.

Could you provide step-by-step instructions for reproducing the crash? I have Directory Opus so I can look into it. (Screenshots or video would be best since describing UI in words is tricky).

**Comment by sschuberth** [comment]
I'll try to come up with a video later, but for me the crash happens when search Postbank PDF account statements for contained text. *Maybe* you can reproduce it by downloading https://github.com/sschuberth/stan/blob/master/lib/src/funTest/assets/PB_KAZ_KtoNr_9999999999_06-04-2017_1200.pdf and https://github.com/sschuberth/stan/blob/master/lib/src/funTest/assets/317970916-PB-KAZ-KtoNr-0914083113-03-06-2016-0313.pdf to a folder, navigating to that folder in Opus, and then use Ctrl-F to search files, and enter some text in the field for searching file content.

**Comment by jupester** [comment]
I have been encountering the same issue, I have been testing all the dailys and pre-releases of 3.3 hoping the issue would get fixed but now since it has hit stable (I tested all recent builds incl 3.3.2), I came to report it and I see this topic.

In my case I am instead using XYplorer as my file manager and Sumatra 32bit plugin.  The steps to reproduce is in XYplorer to goto any folder with PDF's in it and then just enter this text in the addressbar "?cont:the" (without the quotes) which will perform a contents search of the pdfs in the current folder for the word "the", it then crashes and takes down XYplorer with it, and like the above user, using v3.2 is ok.  Also sometimes when having a file selected and toggling the preview pane on it will also crash XYplorer, and it isn't because of the selected PDF file, because it only happens sometimes, even on the same file, so I won't supply a "bad" pdf, yet again using v3.2 instead does not crash XYplorer.

**Comment by kjk** [comment]
How do I setup XYplorer to use Sumatra plugin? How do I know it's Sumatra that's crashing and taking down XYplorer?

**Comment by GitHubRulesOK** [comment]
@kjk 
It should be using explorer functions at present am trying to ensure 3.2 is confirmed as stable but I am on 64 and last post says using 32 bit !!!!

**Comment by jupester** [comment]
> 
> 
> How do I setup XYplorer to use Sumatra plugin? How do I know it's Sumatra that's crashing and taking down XYplorer?

Well I only have Sumatra installed so it automatically uses it, and I use 32bit because XYplorer is 32bit software, I am on 64bit OS though, but if you have multiple 32bit handlers and iFilters installed then you can tell what handler is in use by entering this in the XYplorer addressbar:  `e get("previewhandler", "pdf");` also you can force it to use a particular handler if you want, instructions are here [https://www.xyplorer.com/release_18.00.php#UDPH](https://www.xyplorer.com/release_18.00.php#UDPH)

**Comment by GitHubRulesOK** [comment]
@kjk 
SumatraPDF 3.2 will NOT offer to install preview or search on a 64 bit system
The abuse of 32bit dlls in a 64bit system is the reason XY explorer **should crash.**

DoPus is different it uses 64bit methods on a 64 bit system

**Comment by jupester** [comment]
@GitHubRulesOK: You can extract the dlls and install them manually or run the setup with the commandline switch -with-filter, I wouldn't call that abuse! and using x32 on 64bit OS should work it is not the reason for the crash.

Also I should add that you can actually use the 64bit iFilter if you have that installed anyway, here is where you enable it in the XYplorer:

Configuration | Other | Shell Integration | 64-bit Windows | Use 64-bit IFilters for content search

I haven't tested x64 version though, and my report is about x32 specifically because I would prefer to use the 32bit, because the x64 needs to run a helper process, I assume it will probably still crash using x64 plugin though anyway based on the initial report.

**Comment by GitHubRulesOK** [comment]
32bit filters on 64bit system may sometimes work (but not without issues) and previously declared as unsupported thus no longer available from recent 32bit installer on a 64bit system.
I tried setting 64 bit iFilter on in 32bit XYexplorer and whilst 3.2 64bit preview works the iFilter appears to be non functional using ?cont:the I get nothing whilst MS explorer returns 7 pdfs from over 20,000  (mainly non pdf) files I cancelled at that point, and am reminded why iFilter or similar are not normally installed (background searches thrash productivity)

e get("previewhandler", "pdf"); only returns the one guid (the previewer one)
is there an iFilter command ?

**Comment by jupester** [comment]
No there is no corresponding command within XY for iFilters because the OS only allows 1 to be installed at a time anyway.  It sounds like you don't have it installed properly, you can use this to check though:

[https://www.nirsoft.net/utils/search_filter_view.html](https://www.nirsoft.net/utils/search_filter_view.html)

or manually try registering the iFilter dll again and restarting XYplorer.

**Comment by GitHubRulesOK** [comment]
I did not see your  limit to pdf and on 32 bit where todays preview and ifilter are I do not crash xyexplorer

but on a search  it simply crashed like this
Error:  	6 (0x00000006)
Desc:   	Overflow
Dll:    	0
Proc:   	HasContentsText@0: C:\Users\K\Downloads\test.xps

Source: 	XYplorer
XY ver: 	XYplorer 22.10.0000 (Virgin)
OS:     	Windows 7 Professional (Service Pack 1), 32-bit, Build 7601
Locale: 	2057 (en-GB)
ANSI:   	1252, ACP: 1252  (ANSI - Latin I)
Font:   	Segoe UI 9, DBCS: No
DPI:    	96 (100%), 1440x900

Date:   	2021-07-20 22:28:21
![image](https://user-images.githubusercontent.com/12754913/126398374-f38383fe-b05b-4058-a915-dd58e168efb2.png)

**Comment by jupester** [comment]
I never contents search xps, and because the xps plugin hasn't been updated since Windows 7 especially the x32 version there may be issues with it, I agree.  The crash I mention using Sumatra does not show an error dialog at all it force closes XYplorer silently.

`I did not see your limit to pdf and on 32 bit where todays preview and ifilter are I do not crash xyexplorer`

I don't understand your reply, does that mean you managed to install both iFilter and the Preview Handler and are using the most recent v3.3 - 32bit version and both are definitely functioning correctly? and you have both x64 options disabled now within XYplorer?

Your suggestion to use Acrobat is no-good as it has many issues for me unrelated to this, and the TET filter only works on files upto 10mb, and I wasn't asking for alternatives anyway, I can just continue to use v3.2 Sumatra if your just going to just say what I am doing is unsupported, but it is my opinion there is an issue with the v3.3 plugin whether it is x32 or x64 although as I mentioned I haven't tested x64 yet.  I was just hoping the developer would be willing to try fixing issues in the current version, maybe if he attempts it for the original poster it will fix my issues too, I will wait in hope I guess.

**Comment by GitHubRulesOK** [comment]
I am saying the 3.4.13790 32 bit handlers work perfectly well on 32 bit windows

**Comment by kjk** [comment]
I can reproduce the crash under windbg using latest PdfFilter.dll: `windbg ifilttst.exe /v 1 /i .\bug1972-manual.pdf`

```
00 00007ff8`1e92cde4 : 00007ff8`1e9acdd0 00007ff8`1e991bc8 000002d0`6e8b0000 000002d0`709d9190 : ntdll!RtlpBreakPointHeap+0x16
01 00007ff8`1e8ce115 : 000002d0`6e8b0000 000002d0`709d9190 00000000`00000000 000002d0`6e8b0000 : ntdll!RtlpValidateHeapEntry+0x5eb94
02 00007fff`a1ae750f : 000002d0`6e8b0000 00000000`00000001 00000000`00000000 000002d0`70b7b370 : ntdll!RtlValidateHeap+0x95
*** WARNING: Unable to verify checksum for C:\Users\kjk\AppData\Local\SumatraPDF\PdfFilter.dll
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\Users\kjk\AppData\Local\SumatraPDF\PdfFilter.dll - 
03 00007fff`da0dc648 : 000002d0`70b7b370 000002d0`70b7b370 000000d6`257ffb70 00000000`00000000 : AcLayers!NS_FaultTolerantHeap::APIHook_RtlFreeHeap+0x12f
04 00007fff`da0ca602 : 000002d0`70b7b370 000000d6`00000000 00000000`00000000 00000000`5d1b2867 : PdfFilter!DllUnregisterServer+0x11414

```

**Comment by kjk** [comment]
@sschuberth @jupester I've fixed a crash in PdfFilter.dll Can't say for sure it's your crash.

Please try with latest pre-release (https://www.sumatrapdfreader.org/prerelease, build 13799 or later) and let me know if it fixes the crash you've seen

**Comment by jupester** [comment]
It is fixed for me! Thank you, Thank you, Thank you, I was getting a little discouraged with GitHubRulesOK's replies, but thanks so much for looking into it. 

Also I should add because v3.3.2 stable was just released I had been trying the successive versions of that (3.3, 3.3.1 etc), I didn't think about checking for pre-releases yet, anyway I tried v3.4.13799 and I can't reproduce the crash that occurred sometimes when opening the Preview Pane either, so that looks like it may have been resolved very recently too, but that wasn't the subject of my initial post anyway, the iFilter was mainly.

I have to say what great  software SumatraPDF is and really appreciate it, Thanks again!.

**Comment by GitHubRulesOK** [comment]
For what its worth after my reboot and install latest pdfilter 
xyexplorer 64 is working with 64 bit filter "the" is found in 4 out of 8 pdf out of 10 files

**Comment by kjk** [comment]
released 3.3.3 with the fix

**Comment by sschuberth** [comment]
I also confirm the fix to be working (in Directory Opus 64-bit), thanks a lot!
